dbt_meshify/main.py

```diff
@@ -48,26 +53,68 @@ def operation():
 
 
 @cli.command(name="connect")
-@click.argument("projects-dir", type=click.Path(exists=True), default=".")
-def connect(projects_dir):
+@project_paths
+@projects_dir
+@exclude_projects
+@read_catalog
+def connect(
+    project_paths: tuple, projects_dir: Path, exclude_projects: List[str], read_catalog: bool
+):
     """
-    !!! info
-        This command is not yet implemented
 
     Connects multiple dbt projects together by adding all necessary dbt Mesh constructs
-    """
-    holder = DbtProjectHolder()
-
-    while True:
-        path_string = input("Enter the relative path to a dbt project (enter 'done' to finish): ")
-        if path_string == "done":
-            break
-
-        path = Path(path_string).expanduser().resolve()
-        project = DbtProject.from_directory(path, read_catalog)
-        holder.register_project(project)
 
-    print(holder.project_map())
+    """
+    if project_paths and projects_dir:
+        raise click.BadOptionUsage(
+            option_name="project_paths",
+            message="Cannot specify both project_paths and projects_dir",
+        )
+    # 1. initialize all the projects supplied to the command
+    # 2. compute the dependency graph between each combination of 2 projects in that set
+    # 3. for each dependency, add the necessary dbt Mesh constructs to each project.
+    #    This includes:
+    #    - adding the dependency to the dependencies.yml file of the downstream project
+    #    - adding contracts and public access to the upstream models
+    #    - deleting the source definition of the upstream models in the downstream project
+    #    - updating the `{{ source }}` macro in the downstream project to a {{ ref }} to the upstream project
+
+    linker = Linker()
+    if project_paths:
+        dbt_projects = [
+            DbtProject.from_directory(project_path, read_catalog) for project_path in project_paths
+        ]
+
+    if projects_dir:
+        dbt_project_paths = [path.parent for path in Path(projects_dir).glob("**/dbt_project.yml")]
+        all_dbt_projects = [
+            DbtProject.from_directory(project_path, read_catalog)
+            for project_path in dbt_project_paths
+        ]
+        dbt_projects = [
+            project for project in all_dbt_projects if project.name not in exclude_projects
+        ]
+
+    dbt_project_combinations = [combo for combo in combinations(dbt_projects, 2)]
+    for dbt_project_combo in dbt_project_combinations:
```

Comment:
This will certainly work for now, but will resulting in a combinatorial explosion if we have more than a few projects. In a future issue, I'd recommend we update some of our linker code to be agnostic to the number of projects being evaluated.

Reply:
As an update, this implementation has two rough edge case. 

Suppose we are handling three projects, `A`, `B`, and `C`, such that there exists dependencies `A -> B`  and `A -> C`. Under this logic this script will check for any dependencies between `B` and `C` (which is a good thing!), but will find that both `B` and `C` have a shared dependency ... their common model from project A!

Furthermore, this also shows that the same upstream resource from Project A can be identified as a dependency in `B` and `C` multiple times. This suggests to me that project dependencies should be pooled into a set (for deduplication) prior to evaluation.

By way of an example:
```console
dbt-meshify-py3.11) > $ dbt-meshify connect --project-paths test-projects/source-hack/src_proj_a test-projects/source-hack/src_proj_b test-projects/source-hack/dest_proj_a                                              [±connect-command ●●]
15:52:52 | INFO | Executing dbt parse...
15:52:53 | INFO | Generating catalog with dbt docs generate...
15:52:53 | INFO | Executing dbt parse...
15:52:53 | INFO | Generating catalog with dbt docs generate...
15:52:53 | INFO | Executing dbt parse...
15:52:53 | INFO | Generating catalog with dbt docs generate...
15:52:54 | INFO | Found dependency between src_proj_a and src_proj_b: ProjectDependency(upstream_resource='model.src_proj_a.shared_model', upstream_project_name='src_proj_a', downstream_resource='source.src_proj_b.src_proj_a.shared_model', downstream_project_name='src_proj_b', type=<ProjectDependencyType.Source: 'source'>)
15:52:54 | INFO | Adding public access to shared_model at models/_models.yml
15:52:54 | SUCCESS | Successfully update model access : model.src_proj_a.shared_model is now public
15:52:54 | INFO | Adding contract to shared_model at models/_models.yml
15:52:54 | SUCCESS | Successfully added contract to model: model.src_proj_a.shared_model
15:52:54 | SUCCESS | Successfully replaced source function with ref to upstream resource: source.src_proj_b.src_proj_a.shared_model now calls model.src_proj_a.shared_model directly
15:52:54 | SUCCESS | Successfully deleted unnecessary source: source.src_proj_b.src_proj_a.shared_model
15:52:54 | SUCCESS | Successfully added src_proj_a to src_proj_b's dependencies.yml
15:52:54 | INFO | Found dependency between src_proj_a and dest_proj_a: ProjectDependency(upstream_resource='model.src_proj_a.shared_model', upstream_project_name='src_proj_a', downstream_resource='model.dest_proj_a.downstream_model', downstream_project_name='dest_proj_a', type=<ProjectDependencyType.Package: 'package'>)
15:52:54 | INFO | Adding public access to shared_model at models/_models.yml
15:52:54 | SUCCESS | Successfully update model access : model.src_proj_a.shared_model is now public
15:52:54 | INFO | Adding contract to shared_model at models/_models.yml
15:52:54 | SUCCESS | Successfully added contract to model: model.src_proj_a.shared_model
15:52:54 | SUCCESS | Successfully updated model refs: model.dest_proj_a.downstream_model now references model.src_proj_a.shared_model
15:52:54 | SUCCESS | Successfully added src_proj_a to dest_proj_a's dependencies.yml
15:52:54 | INFO | Found dependency between src_proj_b and dest_proj_a: ProjectDependency(upstream_resource='model.src_proj_a.shared_model', upstream_project_name='dest_proj_a', downstream_resource='source.src_proj_b.src_proj_a.shared_model', downstream_project_name='src_proj_b', type=<ProjectDependencyType.Source: 'source'>)
15:52:54 | INFO | Adding public access to shared_model at models/_models.yml
15:52:54 | SUCCESS | Successfully update model access : model.src_proj_a.shared_model is now public
15:52:54 | INFO | Adding contract to shared_model at models/_models.yml
15:52:54 | SUCCESS | Successfully added contract to model: model.src_proj_a.shared_model
15:52:54 | SUCCESS | Successfully replaced source function with ref to upstream resource: source.src_proj_b.src_proj_a.shared_model now calls model.src_proj_a.shared_model directly
Traceback (most recent call last):
  File "/Users/nicholas/projects/nicholasyager/dbt-meshify-2/dbt_meshify/main.py", line 112, in connect
    linker.resolve_dependency(
  File "/Users/nicholas/projects/nicholasyager/dbt-meshify-2/dbt_meshify/linker.py", line 281, in resolve_dependency
    downstream_editor.update_resource_yml_entry(
  File "/Users/nicholas/projects/nicholasyager/dbt-meshify-2/dbt_meshify/storage/dbt_project_editors.py", line 60, in update_resource_yml_entry
    full_yml_entry = self.file_manager.read_file(current_yml_path)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/nicholas/projects/nicholasyager/dbt-meshify-2/dbt_meshify/storage/file_manager.py", line 63, in read_file
    return yaml.load(full_path.read_text())
                     ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/pathlib.py", line 1058, in read_text
    with self.open(mode='r', encoding=encoding, errors=errors) as f:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.3/Frameworks/Python.framework/Versions/3.11/lib/python3.11/pathlib.py", line 1044, in open
    return io.open(self, mode, buffering, encoding, errors, newline)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: 'test-projects/source-hack/src_proj_b/models/_sources.yml'

15:52:54 | ERROR | Error resolving dependency : ProjectDependency(upstream_resource='model.src_proj_a.shared_model', upstream_project_name='dest_proj_a', downstream_resource='source.src_proj_b.src_proj_a.shared_model', downstream_project_name='src_proj_b', type=<ProjectDependencyType.Source: 'source'>) [Errno 2] No such file or directory: 'test-projects/source-hack/src_proj_b/models/_sources.yml'
```
In this log file, we are seeing an error pop up since the script noticed an extra opportunity to rewrite the source ref in src_proj_b despite 1) it already having been resolved, and 2) it being identified via a non-dependent project!

Reply:
Really appreciate the deep dive here! the solution that you mentioned worked perfectly given the `_hash` method you implemented -- pooling the dependecies into a set for this command ends up with 2 instead of 3 dependecies, which properly resolve: 

```bash
dbt-meshify connect --project-paths test-projects/source-hack/src_proj_a test-projects/source-hack/src_proj_b test-projects/source-hack/dest_proj_a
16:45:31 | INFO | Executing dbt parse...
16:45:31 | INFO | Generating catalog with dbt docs generate...
16:45:32 | INFO | Executing dbt parse...
16:45:32 | INFO | Generating catalog with dbt docs generate...
16:45:32 | INFO | Executing dbt parse...
16:45:32 | INFO | Generating catalog with dbt docs generate...
16:45:33 | INFO | Found 1 dependencies between src_proj_a and src_proj_b
16:45:33 | INFO | Found 1 dependencies between src_proj_a and dest_proj_a
16:45:33 | INFO | Found 1 dependencies between src_proj_b and dest_proj_a
16:45:33 | INFO | Found 2 unique dependencies between all projects.
16:45:33 | INFO | Resolving dependency between model.src_proj_a.shared_model and source.src_proj_b.src_proj_a.shared_model
16:45:33 | INFO | Adding public access to shared_model at models/_models.yml
16:45:33 | SUCCESS | Successfully update model access : model.src_proj_a.shared_model is now public
16:45:33 | INFO | Adding contract to shared_model at models/_models.yml
16:45:33 | SUCCESS | Successfully added contract to model: model.src_proj_a.shared_model
16:45:33 | SUCCESS | Successfully replaced source function with ref to upstream resource: source.src_proj_b.src_proj_a.shared_model now calls model.src_proj_a.shared_model directly
16:45:33 | SUCCESS | Successfully deleted unnecessary source: source.src_proj_b.src_proj_a.shared_model
16:45:33 | SUCCESS | Successfully added src_proj_a to src_proj_b's dependencies.yml
16:45:33 | INFO | Resolving dependency between model.src_proj_a.shared_model and model.dest_proj_a.downstream_model
16:45:33 | INFO | Adding public access to shared_model at models/_models.yml
16:45:33 | SUCCESS | Successfully update model access : model.src_proj_a.shared_model is now public
16:45:33 | INFO | Adding contract to shared_model at models/_models.yml
16:45:33 | SUCCESS | Successfully added contract to model: model.src_proj_a.shared_model
16:45:33 | SUCCESS | Successfully updated model refs: model.dest_proj_a.downstream_model now references model.src_proj_a.shared_model
16:45:33 | SUCCESS | Successfully added src_proj_a to dest_proj_a's dependencies.yml
```